test/cpp/api/transformer.cpp

```diff
@@ -0,0 +1,334 @@
+
+  // key_padding_mask
+  key_padding_mask[0][2] = 1;
+  key_padding_mask[1][1] = 1;
+  key_padding_mask[1][2] = 1;
+  result = model(decoder_input, memory_input, t_mask, m_mask,
+                 key_padding_mask).detach();
+  ref_output = torch::tensor({{{2.430025, 0.027643, -0.601164, -0.073476},
+                               {2.4323, 0.029375, -0.599553, -0.071881}},
+                              {{2.428523, 0.026838, -0.602226, -0.07391},
+                               {2.432634, 0.029842, -0.599318, -0.071253}},
+                              {{2.432278, 0.028152, -0.599555, -0.074139},
+                               {2.432659, 0.029244, -0.599294, -0.072382}}},
+                               tensor_options);
+  ASSERT_EQ(result.sizes().size(),ref_output.sizes().size());
+  ASSERT_TRUE(torch::allclose(result, ref_output, 1e-7, 1e-5,
+                              /*equal_nan=*/true));
+
+  // memory_key_padding_mask
+  at::Tensor t_key_padding_mask = {};
+  key_padding_mask = torch::zeros({2, 5}) == 1;
+  result = model(decoder_input, memory_input, t_mask, m_mask,
+                 t_key_padding_mask, key_padding_mask).detach();
+  ref_output = torch::tensor({{{2.430065, 0.027862, -0.601136, -0.073096},
+                               {2.431935, 0.028907, -0.599809, -0.072488}},
+                              {{2.428457, 0.027053, -0.602275, -0.073462},
+                               {2.431970, 0.029387, -0.599789, -0.071621}},
+                              {{2.431934, 0.028196, -0.599802, -0.073809},
+                               {2.432306, 0.028858, -0.599542, -0.072846}}});
+  ASSERT_EQ(result.sizes().size(),ref_output.sizes().size());
+  ASSERT_TRUE(torch::allclose(result, ref_output, 1e-7, 1e-5,
+                              /*equal_nan=*/true));
+
+  // memory_key_padding_mask
+  key_padding_mask[0][4] = 1;
+  key_padding_mask[1][3] = 1;
+  key_padding_mask[1][4] = 1;
+  result = model(decoder_input, memory_input, t_mask, m_mask,
+                 t_key_padding_mask, key_padding_mask).detach();
+  ref_output = torch::tensor({{{2.429757, 0.027358, -0.601351, -0.073816},
+                               {2.432692, 0.028583, -0.599263, -0.073634}},
+                              {{2.428247, 0.02662, -0.602419, -0.074123},
+                               {2.432657, 0.029055, -0.599293, -0.072732}},
+                              {{2.431515, 0.027687, -0.600096, -0.074459},
+                               {2.433075, 0.028543, -0.598987, -0.073985}}},
+                               tensor_options);
+  ASSERT_EQ(result.sizes().size(),ref_output.sizes().size());
+  ASSERT_TRUE(torch::allclose(result, ref_output, 1e-7, 1e-5,
+                              /*equal_nan=*/true));
+
+}
+
+TEST_F(TransformerTest, TransformerDecoderLayer){
+  transformer_decoder_layer_test_helper(false);
+}
+
+TEST_F(TransformerTest, TransformerDecoderLayerCUDA){
+  // Check to prevent execution on non CUDA devices
+  if (torch::cuda::is_available()){
```

Comment:
As long as you change the test name to xxxx_CUDA, you don't need to check cuda available, our frame will only call this function if cuda is available